src/ui/public/saved_objects/saved_objects_client.js

```diff
@@ -0,0 +1,102 @@
+import { resolve as resolveUrl, format as formatUrl } from 'url';
+import { pick, partial } from 'lodash';
+
+import { SavedObject } from './saved_object';
+
+const join = (...uriComponents) => (
+  uriComponents.filter(Boolean).map(encodeURIComponent).join('/')
+);
+
+export class SavedObjectsClient {
+  constructor($http, basePath, PromiseCtor = Promise) {
+    this._$http = $http;
+    this._apiBaseUrl = `${basePath}/api/saved_objects/`;
+    this._PromiseCtor = PromiseCtor;
+  }
+
+  get(type, id) {
+    if (!type || !id) {
+      return this._PromiseCtor.reject(new Error('requires type and id'));
+    }
+
+    return this._request('GET', this._getUrl([type, id])).then(resp => {
+      return new this.ObjectClass(resp);
+    });
+  }
+
+  delete(type, id) {
+    if (!type || !id) {
+      return this._PromiseCtor.reject(new Error('requires type and id'));
+    }
+
+    return this._request('DELETE', this._getUrl([type, id]));
+  }
+
+  update(type, id, attributes, { version } = {}) {
+    if (!type || !id || !attributes) {
+      return this._PromiseCtor.reject(new Error('requires type, id and attributes'));
+    }
+
+    const body = {
+      attributes,
+      version
+    };
+
+    return this._request('PUT', this._getUrl([type, id]), body);
+  }
+
+  create(type, body) {
+    if (!type || !body) {
+      return this._PromiseCtor.reject(new Error('requires type and body'));
+    }
+
+    const url = this._getUrl([type]);
+
+    return this._request('POST', url, body);
+  }
+
+  find(options = {}) {
+    const url = this._getUrl([options.type]);
+    const validOptions = pick(options, ['from', 'size', 'fields', 'filter']);
+
+    return this._request('GET', url, validOptions).then(resp => {
+      resp.data = resp.data.map(d => new this.ObjectClass(d));
+      return resp;
+    });
+  }
+
+  get ObjectClass() {
+    return partial(SavedObject, this);
+  }
+
+  _getUrl(path, query) {
+    if (!path && !query) {
+      return this._apiBaseUrl;
+    }
+
+    return resolveUrl(this._apiBaseUrl, formatUrl({
+      pathname: join(...path),
+      query: pick(query, value => value != null)
+    }));
+  }
+
+  _request(method, url, body) {
+    const options = { method, url };
+
+    if (method === 'GET' && body) {
```

Comment:
Arguments having different meaning based on other arguments is pretty confusing, and this is unnecessary since the second argument to `SavedObjectsClient#_getUrl()` already allows setting query string parameters. This should probably just throw an error like "body is not allowed on `GET` requests"